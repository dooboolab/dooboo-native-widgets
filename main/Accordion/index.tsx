import React, { FC } from 'react';
import AccrordionItem from './AccordionItem';
import styled from 'styled-components/native';

const Container = styled.ScrollView`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

type ItemType = {
  itemTitle: string;
  itemBodies: Array<string>;
}

interface Props {
  data: Array<ItemType>;
  isAnimated?: boolean;
  collapsedWhenRedered: boolean;
  animDuration?: number;
  activeOpacity?: number;
}

const Accordion: FC<Props> = (props) => {
  const {
    data,
    isAnimated,
    collapsedWhenRedered,
    animDuration,
    activeOpacity,
  } = props;

  return (
    <Container>
      {
        data.map((itemData, idx) => {
          return (
            <AccrordionItem
              key={idx}
              itemData={itemData}
              isAnimated={isAnimated}
              collapsedWhenRedered={collapsedWhenRedered}
              animDuration={animDuration}
              activeOpacity={activeOpacity}
            />
          );
        })
      }
    </Container>
  );
};

export default Accordion;
