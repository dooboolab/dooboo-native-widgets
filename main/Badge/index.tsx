import React, { FC } from 'react';

import { StyleSheet } from 'react-native';
import styled from 'styled-components/native';

interface StyleProps {
  color?: string;
  opacity?: number;
  badgePlacement?: 'left' | 'top' | 'bottom' | 'right';
  border?: string;
  textColor?: string;
}

export interface BadgeProps extends StyleProps{
  count?: number;
  maximumCount?: number;
  showZero?: boolean;
  opacityVisible?: boolean;
  variant?: string;
}

const StyledView = styled.View<StyleProps>`
  position: absolute;
  top: -15px;
  ${(props) => props.badgePlacement}: -10px;
  width: auto;
  min-width: 45px;
  height: 45px;
  border-color: ${(props) =>
    props.border ? props.border : '#00ff0000'};
  border-width: 3px;
  background-color: ${(props) => props.color};
  border-radius: 50px;
  justify-content: center;
  align-items: center;
  opacity: ${(props) => props.opacity};
`;

const StyledText = styled.Text<StyleProps>`
  color: ${(props) => props.textColor};
  text-align: center;
  padding: 5px;
  margin-left: 3px;
  margin-right: 3px;
`;

const StyledDotView = styled.View<StyleProps>`
  position: absolute;
  top: -5px;
  ${(props) => props.badgePlacement}: -5px;
  width: 20px;
  height: 20px;
  background-color: ${(props) => props.color};
  border-radius: 50px;
  justify-content: center;
  align-items: center;
`;

const styles = StyleSheet.create({
  fontDesign: {
    fontWeight: '500',
    fontSize: 20,
  },
});

const Badge: FC<BadgeProps> = (props) => {
  const {
    count = 1,
    color = '#34AFF9',
    maximumCount = 99,
    showZero,
    opacityVisible = true,
    variant = 'standard',
    badgePlacement: position = 'right',
    border,
    textColor = '#FFFFFF',
  } = props;

  if (!showZero) {
    if (count === 0) return null;
  }

  if (variant === 'dot') {
    return <StyledDotView color={color} badgePlacement={position} />;
  } else {
    const opacity = (maximumCount < count && opacityVisible) ? 0.6 : 1;
    const countText = (maximumCount < count) ? count + '+' : count;

    return (
      <StyledView
        color={color}
        opacity={opacity}
        badgePlacement={position}
        border={border}>
        <StyledText textColor={textColor} style={styles.fontDesign}>
          {countText}
        </StyledText>
      </StyledView>
    );
  }
};

export { Badge };
