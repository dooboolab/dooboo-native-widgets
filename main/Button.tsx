import {
  ActivityIndicator,
  LayoutRectangle,
  StyleProp,
  StyleSheet,
  Text,
  TextProps,
  TouchableOpacity,
  TouchableOpacityProps,
  View,
  ViewStyle,
} from 'react-native';
import React, {useRef, useState} from 'react';

import {useHover} from 'react-native-web-hooks';

const defaultStyles = StyleSheet.create({
  container: {
    alignSelf: 'stretch',
    paddingHorizontal: 12,
    paddingVertical: 8,
    borderRadius: 6,

    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
  },
  text: {
    fontSize: 14,
    color: '#069ccd',
  },
  disabledButton: {
    backgroundColor: '#cccccc',
    borderColor: 'rgb(200, 200, 200)',
  },
  disabledText: {
    color: '#969696',
  },
  hovered: {
    shadowColor: 'black',
    shadowOffset: {
      width: 0,
      height: 4,
    },
    shadowOpacity: 0.24,
    shadowRadius: 16,
    elevation: 10,
  },
});

type StylesType = Partial<StyleSheet.NamedStyles<typeof defaultStyles>>;

interface Props {
  testID?: string;
  indicatorColor?: string;
  loading?: boolean;
  disabled?: boolean;
  style?: StyleProp<ViewStyle>;
  styles?: StylesType;
  leftElement?: React.ReactElement;
  rightElement?: React.ReactElement;
  activeOpacity?: TouchableOpacityProps['activeOpacity'];
  text?: string;
  onPress?: TouchableOpacityProps['onPress'];
  touchableOpacityProps?: Partial<TouchableOpacityProps>;
  textProps?: Partial<TextProps>;
}

function Button({
  testID,
  disabled,
  loading,
  style,
  styles,
  indicatorColor = '#ffffff',
  leftElement,
  rightElement,
  activeOpacity = 0.7,
  text,
  onPress,
  touchableOpacityProps,
  textProps,
}: Props): React.ReactElement {
  const ref = useRef<TouchableOpacity>(null);
  const hovered = useHover(ref);
  const [layout, setLayout] = useState<LayoutRectangle>();

  return (
    <TouchableOpacity
      testID={testID}
      ref={ref}
      activeOpacity={activeOpacity}
      onPress={onPress}
      delayPressIn={50}
      disabled={disabled}
      style={style}
      {...touchableOpacityProps}>
      {loading ? (
        <View
          style={[
            defaultStyles.container,
            styles?.container,
            {
              width: layout?.width,
              height: layout?.height,
            },
            hovered && !disabled && [defaultStyles.hovered, styles?.hovered],
            disabled && [defaultStyles.disabledButton, styles?.disabledButton],
          ]}>
          <ActivityIndicator size="small" color={indicatorColor} />
        </View>
      ) : (
        <View
          style={[
            defaultStyles.container,
            styles?.container,
            hovered && !disabled && [defaultStyles.hovered, styles?.hovered],
            disabled && [defaultStyles.disabledButton, styles?.disabledButton],
          ]}
          onLayout={(e) => setLayout(e.nativeEvent.layout)}>
          {leftElement}
          <Text
            style={[
              defaultStyles.text,
              styles?.text,
              disabled && [defaultStyles.disabledText, styles?.disabledText],
            ]}
            {...textProps}>
            {text}
          </Text>
          {rightElement}
        </View>
      )}
    </TouchableOpacity>
  );
}

export {Button};
