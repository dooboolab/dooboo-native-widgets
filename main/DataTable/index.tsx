import * as React from 'react';

import {
  ScrollView,
  StyleProp,
  StyleSheet,
  Text,
  View,
  ViewStyle,
} from 'react-native';

import CheckBox from './CheckBox';
import DataTableCell from './DataTableCell';
import DataTableHeader from './DataTableHeader';
import DataTableRow from './DataTableRow';
import DataTableTitle from './DataTableTitle';
import { isHasValue } from './utils';
import styled from 'styled-components/native';

export interface Props {
  style?: StyleProp<ViewStyle>;
  data: Array<Record<string, any>>;
  isCheckAble?: boolean;
  customGroup?: Array<string>;
}

type DataTableNamespace = {
  Title: typeof DataTableTitle;
  Header: typeof DataTableHeader;
  Row: typeof DataTableRow;
  Cell: typeof DataTableCell;
};

const Container = styled.View`
  flex: 1;
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
`;

const DataTable: React.FC<Props> & DataTableNamespace = ({
  data,
  isCheckAble,
  customGroup,
  style,
  ...rest
}) => {
  /** default Data key */
  const group = data.reduce((acc, current) => {
    Array.from(Object.assign(acc, Object.keys(current)));
    return acc;
  }, []);

  /** checking interaction */
  const [selected, setSelected] = React.useState<string[]>([]);
  const isSelected = (name: string): boolean => selected.indexOf(name) !== -1;
  const handleClick = (name: string): void => {
    const selectedIndex = selected.indexOf(name);
    let newSelected: string[] = [];
    if (selectedIndex === -1) {
      newSelected = newSelected.concat(selected, name);
    } else if (selectedIndex === 0) {
      newSelected = newSelected.concat(selected.slice(1));
    } else if (selectedIndex === selected.length - 1) {
      newSelected = newSelected.concat(selected.slice(0, -1));
    } else if (selectedIndex > 0) {
      newSelected = newSelected.concat(
        selected.slice(0, selectedIndex),
        selected.slice(selectedIndex + 1),
      );
    }
    setSelected(newSelected);
  };
  /** Render */
  return (
    <ScrollView
      style={{
        width: '100%',
        height: '100%',
      }}>
      <ScrollView horizontal>
        <Container {...rest} style={[style]}>
          <View>
            <Text>
              {selected.length > 0 ? `${selected.length} item selected` : null}{' '}
            </Text>
          </View>
          <DataTable.Header style={{ backgroundColor: '#f9f9f9' }}>
            <DataTable.Title style={[!isCheckAble && styles.short]} />

            {/** have a customGroup or undefined  */}
            {(customGroup || group)?.map((field: any, index: number) => {
              return (
                <DataTable.Title
                  numberOfLines={1}
                  key={`${field}-${index}`}
                  style={[styles.center, index === 0 ? styles.short : null]}>
                  {field}
                </DataTable.Title>
              );
            })}
          </DataTable.Header>
          {data.map((item, i) => {
            const isItemSelected = isSelected(item[group[0]]);
            return (
              <DataTable.Row
                key={`row-${item}-${i}`}
                isChecked={!!isItemSelected}
                style={{ backgroundColor: 'white' }}>
                {/* If CheckAble is true */}
                {isCheckAble ? (
                  <DataTable.Cell style={[{ justifyContent: 'center' }]}>
                    <CheckBox
                      onClick={(): void => handleClick(item[group[0]])}
                      checked={!!isItemSelected}
                    />
                  </DataTable.Cell>
                ) : (
                  <DataTable.Cell style={[!isCheckAble && styles.short]} />
                )}
                {/** Body */}
                {group?.map((param: any, index: number) => {
                  return (
                    <DataTable.Cell
                      key={`cell-${param}-${index}`}
                      style={[
                        styles.center,
                        index === 0 ? styles.short : null,
                      ]}>
                      {isHasValue(item[param])}
                    </DataTable.Cell>
                  );
                })}
              </DataTable.Row>
            );
          })}
        </Container>
      </ScrollView>
    </ScrollView>
  );
};

DataTable.Title = DataTableTitle;
DataTable.Header = DataTableHeader;
DataTable.Row = DataTableRow;
DataTable.Cell = DataTableCell;

const styles = StyleSheet.create({
  short: {
    width: 50,
  },
  center: {
    justifyContent: 'center',
  },
});

export default DataTable;
