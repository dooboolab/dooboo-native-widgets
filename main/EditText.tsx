import {Platform, StyleSheet, Text, TextInput, View} from 'react-native';
import React, {FC, useState} from 'react';
import type {StyleProp, TextInputProps, ViewStyle} from 'react-native';

const defaultStyle: ViewStyle = {
  alignSelf: 'stretch',
  flex: 1,

  flexDirection: 'column',
};

const defaultRowStyles = StyleSheet.create({
  container: {
    alignSelf: 'stretch',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    borderBottomWidth: 0.3,
    borderBottomColor: '#666666',
  },
  labelText: {
    fontSize: 14,
    color: '#b9b9c4',
    width: 100,
  },
  input: {
    paddingVertical: 12,
    paddingHorizontal: 8,
    fontSize: 14,
    fontWeight: 'bold',
    flex: 1,
    color: '#2c374e',
  },
  errorText: {
    marginTop: 4,
    fontSize: 12,
    color: '#991111',
  },
});

const defaultColumnStyles = StyleSheet.create({
  container: {
    alignSelf: 'stretch',
    justifyContent: 'space-between',
    borderBottomWidth: 0.3,
    borderBottomColor: '#666666',

    flexDirection: 'column',
  },
  labelText: {
    fontSize: 14,
    color: '#b9b9c4',
  },
  input: {
    paddingVertical: 12,
    fontSize: 14,
    fontWeight: 'bold',
    flex: 1,
    color: '#2c374e',
  },
  errorText: {
    marginTop: 8,
    fontSize: 12,
    color: '#991111',
  },
});

type StylesType = Partial<StyleSheet.NamedStyles<typeof defaultRowStyles>>;

type Props = {
  testID?: TextInputProps['testID'];
  textInputProps?: TextInputProps;
  style?: StyleProp<ViewStyle>;
  styles?: StylesType;
  labelText?: string;
  errorText?: string;
  value?: TextInputProps['value'];
  onChange?: TextInputProps['onChange'];
  onChangeText?: TextInputProps['onChangeText'];
  placeholder?: TextInputProps['placeholder'];
  placeholderTextColor?: TextInputProps['placeholderTextColor'];
  onFocus?: TextInputProps['onFocus'];
  onBlur?: TextInputProps['onBlur'];
  editable?: TextInputProps['editable'];
  autoCapitalize?: TextInputProps['autoCapitalize'];
  secureTextEntry?: TextInputProps['secureTextEntry'];
  onSubmitEditing?: TextInputProps['onSubmitEditing'];
  focusColor?: string;
  errorColor?: string;
  disableColor?: string;
  type?: 'row' | 'column';
};

const EditText: FC<Props> = ({
  testID,
  textInputProps,
  style,
  styles,
  labelText = '',
  errorText = '',
  value = '',
  placeholder,
  placeholderTextColor,
  onChange,
  onChangeText,
  onFocus,
  onBlur,
  autoCapitalize = 'none',
  secureTextEntry = false,
  editable = true,
  focusColor = '#9A77FF',
  errorColor = '#ff7676',
  disableColor = '#999999',
  type = 'column',
}) => {
  const [focused, setFocused] = useState(false);

  const defaultStyles = type === 'row' ? defaultRowStyles : defaultColumnStyles;

  return (
    <View style={[defaultStyle, style]}>
      <View
        style={[
          defaultStyles.container,
          {
            borderColor: errorText
              ? errorColor
              : focused
              ? focusColor
              : disableColor,
            borderBottomColor: errorText
              ? errorColor
              : focused
              ? focusColor
              : disableColor,
          },
        ]}>
        {/* {labelText ? ( */}
        <Text
          style={[
            defaultStyles.labelText,
            styles?.labelText,
            {
              color: errorText
                ? errorColor
                : focused
                ? focusColor
                : disableColor,
            },
          ]}>
          {labelText}
        </Text>
        {/* ) : null} */}
        <TextInput
          {...textInputProps}
          testID={testID}
          autoCapitalize={autoCapitalize}
          secureTextEntry={secureTextEntry}
          style={[
            defaultStyles.input,
            styles?.input,
            // @ts-ignore
            Platform.OS === 'web' && {outlineWidth: 0},
          ]}
          editable={editable}
          onFocus={(e) => {
            setFocused(true);
            onFocus?.(e);
          }}
          onBlur={(e) => {
            setFocused(false);
            onBlur?.(e);
          }}
          value={value}
          placeholder={placeholder}
          placeholderTextColor={placeholderTextColor}
          onChange={onChange}
          onChangeText={onChangeText}
        />
      </View>
      {errorText ? (
        <Text
          style={[
            defaultStyles.errorText,
            styles?.errorText,
            {color: errorColor},
          ]}>
          {errorText}
        </Text>
      ) : null}
    </View>
  );
};

export {EditText};
