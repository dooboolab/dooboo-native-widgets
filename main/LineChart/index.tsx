/* eslint-disable sort-imports */
import React, { FC } from 'react';
import styled from 'styled-components/native';
import { Circle, G, Line, Svg, Text } from 'react-native-svg';
import * as d3 from 'd3';

import { LineChartProps } from './types';

// Styled component declaration
const Container = styled.View`
  width: 100%;
  height: 320;
  background-color: 'transparent';
`;
const HeaderContainer = styled.View`
  flex-direction: row;
  justify-content: center;
  align-items: center;
`;
const ChartContainer = styled.View`
  flex: 1;
  flex-direction: row;
`;
const GraphWrapper = styled.View`
  flex: 1;
  flex-direction: column;
`;

const LineChart: FC<LineChartProps> = (props) => {
  // Destructuring props & Declaring default values
  const {
    screenHeight,
    screenWidth,
    /* ====== [REQUIRED] ====== */
    data = [],
    xAxisKey = 'key3',
    yAxisKey = 'key6',
    yUnit = '100',
    /* ====== [OPTIONAL] ====== */
    header = undefined,
    graphStyle = {
      withLine: true,
      lineColor: '#000000',
      lineWidth: 2,
      withDots: true,
      dotColor: '#ffffff',
      dotStrokeColor: '#000000',
      dotStrokeWidth: 2,
      withText: true,
      textColor: '#000000',
      textStrokeColor: 'none',
      fontSize: '12',
      fontWeight: 'bold',
    },

    /* **** [X-AXIS PARAMETERS] ====== **** */
    xStyle = {
      withLabel: true,
      withIndicator: true,
      withLine: true,
      lineColor: '#000000',
      lineStrokeWidth: 1,
      withText: true,
      textColor: '#000000',
      textStrokeColor: 'none',
      fontSize: '12',
      fontWeight: 'bold',
    },
    /* **** [Y-AXIS PARAMETERS] **** */
    yStyle = {
      withLabel: true,
      withIndicator: true,
      withLine: true,
      lineColor: '#000000',
      lineStrokeWidth: 1,
      withText: true,
      textColor: '#000000',
      textStrokeColor: 'none',
      fontSize: '12',
      fontWeight: 'normal',
    },
  } = props;
  // Initialize
  const dataLen = data.length;
  const dataWithIdx = data.map((item, index) => ({
    ...item,
    idx: dataLen - index,
  }));
  const SVGHeight = screenHeight * 0.2;
  const SVGWidth = screenWidth * 0.75;

  // X-AXIS
  const xDomain = dataWithIdx.map((item) => item.idx);
  const xRange = [0, SVGWidth];
  const xAxis = d3.scalePoint().domain(xDomain).range(xRange);

  // Y-AXIS
  const yDomain = [0, d3.max(data, (item) => item[yAxisKey])];
  const yRange = [0, SVGHeight];
  const yAxis = d3.scaleLinear().domain(yDomain).range(yRange);

  // get Y axis values range
  const getYmaxRange = (data: any, yUnit: string): Array<number> => {
    const range: number[] = [];
    const max =
      Math.trunc(
        Math.max(...data.map((current) => current[yAxisKey]), 0) /
          Number(yUnit),
      ) *
        Number(yUnit) +
      Number(yUnit);
    for (let i = 0; i <= max; i += Number(yUnit)) {
      i <= max && range.push(i);
    }
    return range;
  };

  return (
    <Container>
      {/* Header */}
      <HeaderContainer>{header}</HeaderContainer>
      {/* Graph view */}
      <ChartContainer>
        {/* Graph Y-axis labels view */}
        <Svg height="100%" width={50}>
          {yStyle.withLabel && (
            <G x={24} y={200}>
              {yStyle.withLine && (
                <Line
                  x1={25}
                  y1={25}
                  x2={25}
                  y2={-170}
                  stroke={yStyle.lineColor}
                  strokeWidth={yStyle.lineStrokeWidth}
                />
              )}
              {getYmaxRange(data, yUnit).map((unit, index) => {
                return (
                  <>
                    {yStyle.withIndicator && (
                      <Line
                        key={index}
                        x1={20}
                        y1={yAxis(unit) * -1 + 20}
                        x2={30}
                        y2={yAxis(unit) * -1 + 20}
                        stroke={yStyle.lineColor}
                        strokeWidth={yStyle.lineStrokeWidth}
                      />
                    )}
                    {yStyle.withText && (
                      <Text
                        fill={yStyle.textColor}
                        stroke={yStyle.textStrokeColor}
                        fontSize={yStyle.fontSize}
                        fontWeight={yStyle.fontWeight}
                        x={5}
                        y={yAxis(unit) * -1 + 24}
                        textAnchor="middle">
                        {unit}
                      </Text>
                    )}
                  </>
                );
              })}
            </G>
          )}
        </Svg>
        <GraphWrapper style={{ width: 'auto', height: 'auto' }}>
          {/* Graph X-axis labels view */}
          <Svg id={'graph-svg'} height="100%" width={screenWidth * 0.9}>
            <G x={10} y={210}>
              {/* Graph X-axis labels view */}
              {xStyle.withLabel && xStyle.withLine && (
                <Line
                  x1={xAxis(dataLen) - 30}
                  y1={10}
                  x2={xAxis(dataLen - dataLen + 1) + 35}
                  y2={10}
                  stroke={xStyle.lineColor}
                  strokeWidth={xStyle.lineStrokeWidth}
                />
              )}
              {xStyle.withLabel &&
                dataWithIdx.map((item, index) => {
                  return (
                    <>
                      {xStyle.withIndicator && (
                        <Line
                          x1={xAxis(dataLen - index) + 10}
                          y1={10}
                          x2={xAxis(dataLen - index) + 10}
                          y2={15}
                          stroke={xStyle.lineColor}
                          strokeWidth={xStyle.lineStrokeWidth}
                        />
                      )}
                      {xStyle.withText && (
                        <Text
                          fill={xStyle.textColor}
                          stroke={xStyle.textStrokeColor}
                          fontSize={xStyle.fontSize}
                          fontWeight={xStyle.fontWeight}
                          x={xAxis(dataLen - index) + 10}
                          y={30}
                          textAnchor="middle">
                          {item[xAxisKey]}
                        </Text>
                      )}
                    </>
                  );
                })}
              {/* Graph: Text, Dots & Line */}
              {dataWithIdx.map((item, index) => {
                const xStart = dataLen - index;
                const yStart = item[yAxisKey];
                let xEnd: number;
                let yEnd: number;

                if (index === dataLen - 1) {
                  xEnd = xStart;
                  yEnd = yStart;
                } else {
                  xEnd = dataLen - index - 1;
                  yEnd = dataWithIdx[index + 1][yAxisKey];
                }
                return (
                  <>
                    {graphStyle.withLine && (
                      <Line
                        key={index}
                        x1={xAxis(xStart) + 10}
                        y1={yAxis(yStart) * -1 + 10}
                        x2={xAxis(xEnd) + 10}
                        y2={yAxis(yEnd) * -1 + 10}
                        stroke={graphStyle.lineColor}
                        strokeWidth={graphStyle.lineWidth}
                      />
                    )}
                    {graphStyle.withText && (
                      <Text
                        fill={graphStyle.textColor}
                        stroke={graphStyle.textStrokeColor}
                        fontSize={graphStyle.fontSize}
                        fontWeight={graphStyle.fontWeight}
                        x={xAxis(dataLen - index) + 10}
                        y={-yAxis(item[yAxisKey])}
                        textAnchor="middle">
                        {item[yAxisKey]}
                      </Text>
                    )}
                    {graphStyle.withDots && (
                      <Circle
                        cx={xAxis(dataLen - index) + 10}
                        cy={yAxis(item[yAxisKey]) * -1 + 10}
                        r="4"
                        fill={graphStyle.dotColor}
                        stroke={graphStyle.dotStrokeColor}
                        strokeWidth={graphStyle.dotStrokeWidth}
                      />
                    )}
                  </>
                );
              })}
            </G>
          </Svg>
        </GraphWrapper>
      </ChartContainer>
    </Container>
  );
};

export default LineChart;
