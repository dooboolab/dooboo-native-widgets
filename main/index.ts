export * from './Accordion';
export * from './Button';
export * from './ButtonGroup';
export * from './EditText';
export * from './LoadingIndicator';
export * from './Slider';
export * from './SwitchToggle';
export * from './Select';
