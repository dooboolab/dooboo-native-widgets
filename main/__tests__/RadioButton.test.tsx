import * as React from 'react';

import { RadioButton } from '../../main';
// Note: test renderer must be required after react-native.
import { TouchableOpacity } from 'react-native';
import renderer from 'react-test-renderer';

/**
 * using for animation easing
 * check: https://stackoverflow.com/questions/50793885/referenceerror-you-are-trying-to-import-a-file-after-the-jest-environment-has
 */
jest.useFakeTimers();

// eslint-disable-next-line @typescript-eslint/no-explicit-any
let props: any;
// let component: React.ReactElement;
// let testingLib: RenderResult;

const createTestProps = (obj: Record<string, unknown>): Record<string, unknown> => ({
  navigation: {
    navigate: jest.fn(),
  },
  ...obj,
});

const component = (props?): React.ReactElement => {
  return <RadioButton {...props} />;
};

describe('[RadioButton] render', () => {
  beforeEach(() => {
    props = createTestProps({});
  });

  it('renders without crashing', () => {
    const rendered: renderer.ReactTestRendererJSON | null = renderer
      .create(component(props))
      .toJSON();
    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });

  describe('[RadioButton] Interaction', (): void => {
    it('should simulate onPress', (): void => {
      const handlePress = jest.fn();
      const rendered = renderer.create(
        component({
          testID: 'RADIOBUTTON_ID',
          onPress: handlePress,
        }),
      );
      const radioButtonClick = rendered.root.findByType(TouchableOpacity);
      renderer.act(() => {
        radioButtonClick.props.onPress();
      });
      expect(handlePress).toHaveBeenCalled();
    });
    it('should simulate props', (): void => {
      const rendered = renderer.create(
        component({
          testID: 'RADIOBUTTON_ID',
        }),
      );

      rendered.update(component({ labelPlacement: 'top' }));
      expect(rendered).toMatchSnapshot();
      expect(rendered).toBeTruthy();

      rendered.update(component({ disabled: true }));
      expect(rendered).toMatchSnapshot();
      expect(rendered).toBeTruthy();

      rendered.update(
        component({ disabled: false, value: 0, selectedValue: 1 }),
      );
      expect(rendered).toMatchSnapshot();
      expect(rendered).toBeTruthy();
    });
  });
});
