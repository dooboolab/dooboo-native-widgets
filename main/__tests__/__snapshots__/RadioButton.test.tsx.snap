// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[RadioButton] render [RadioButton] Interaction should simulate props 1`] = `
<
  activeOpacity={1}
  disabled={false}
  isLabelColumn={true}
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "column",
        "justifyContent": "center",
        "paddingBottom": 8,
        "paddingLeft": 0,
        "paddingRight": 0,
        "paddingTop": 8,
      },
    ]
  }
>
  <Text
    disabled={false}
    style={
      Array [
        Object {
          "color": "#000000",
        },
      ]
    }
  >
    
  </Text>
  <View
    borderRadius={12}
    borderWidth={2.4}
    color="rgba(0, 0, 0, 0.54)"
    innerSize={16}
    size={24}
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderColor": "rgba(0, 0, 0, 0.54)",
          "borderRadius": 12,
          "borderWidth": 2.4,
          "height": 24,
          "justifyContent": "center",
          "marginBottom": 7,
          "marginLeft": 7,
          "marginRight": 7,
          "marginTop": 7,
          "width": 24,
        },
        false,
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.54)",
          "borderRadius": 12,
          "height": 16,
          "width": 16,
        }
      }
    />
  </View>
</>
`;

exports[`[RadioButton] render [RadioButton] Interaction should simulate props 2`] = `
<
  activeOpacity={1}
  disabled={true}
  isLabelColumn={false}
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "flex-start",
        "paddingBottom": 8,
        "paddingLeft": 0,
        "paddingRight": 0,
        "paddingTop": 8,
      },
    ]
  }
>
  <View
    borderRadius={12}
    borderWidth={2.4}
    color="#E0E0E0"
    innerSize={16}
    size={24}
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderColor": "#E0E0E0",
          "borderRadius": 12,
          "borderWidth": 2.4,
          "height": 24,
          "justifyContent": "center",
          "marginBottom": 7,
          "marginLeft": 7,
          "marginRight": 7,
          "marginTop": 7,
          "width": 24,
        },
        Object {
          "backgroundColor": "#E0E0E0",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#E0E0E0",
          "borderRadius": 12,
          "height": 16,
          "width": 16,
        }
      }
    />
  </View>
  <Text
    disabled={true}
    style={
      Array [
        Object {
          "color": "#969696",
        },
      ]
    }
  >
    
  </Text>
</>
`;

exports[`[RadioButton] render [RadioButton] Interaction should simulate props 3`] = `
<
  activeOpacity={1}
  disabled={false}
  isLabelColumn={false}
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "flex-start",
        "paddingBottom": 8,
        "paddingLeft": 0,
        "paddingRight": 0,
        "paddingTop": 8,
      },
    ]
  }
>
  <View
    borderRadius={12}
    borderWidth={2.4}
    color="rgba(0, 0, 0, 0.54)"
    innerSize={16}
    size={24}
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderColor": "rgba(0, 0, 0, 0.54)",
          "borderRadius": 12,
          "borderWidth": 2.4,
          "height": 24,
          "justifyContent": "center",
          "marginBottom": 7,
          "marginLeft": 7,
          "marginRight": 7,
          "marginTop": 7,
          "width": 24,
        },
        false,
      ]
    }
  />
  <Text
    disabled={false}
    style={
      Array [
        Object {
          "color": "#000000",
        },
      ]
    }
  >
    
  </Text>
</>
`;

exports[`[RadioButton] render renders without crashing 1`] = `
<
  activeOpacity={1}
  disabled={false}
  isLabelColumn={false}
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "flex-start",
        "paddingBottom": 8,
        "paddingLeft": 0,
        "paddingRight": 0,
        "paddingTop": 8,
      },
    ]
  }
>
  <View
    borderRadius={12}
    borderWidth={2.4}
    color="rgba(0, 0, 0, 0.54)"
    innerSize={16}
    size={24}
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderColor": "rgba(0, 0, 0, 0.54)",
          "borderRadius": 12,
          "borderWidth": 2.4,
          "height": 24,
          "justifyContent": "center",
          "marginBottom": 7,
          "marginLeft": 7,
          "marginRight": 7,
          "marginTop": 7,
          "width": 24,
        },
        false,
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.54)",
          "borderRadius": 12,
          "height": 16,
          "width": 16,
        }
      }
    />
  </View>
  <Text
    disabled={false}
    style={
      Array [
        Object {
          "color": "#000000",
        },
      ]
    }
  >
    
  </Text>
</>
`;
